provisioning/device/devdoc/transport_state_machine_requirements.md

```diff
@@ -0,0 +1,81 @@
+
+
+### connect(callback: (err?: Error) => void): void
+Connect is used to connect the transport
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_002: [** `connect` shall call `_doConnectForFsm`. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_003: [** If `_doConnectForFsm` fails, then `connect` shall fail. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_004: [** If the transport is already connected, then `connect` shall do nothing and call the `callback` immediately. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_005: [** If `connect` is called while a connection is in progress, it shall wait for that connection to complete before calling `callback`. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_006: [** If `connect` is called while the transport is executing the first registration request, it shall do nothing and call `callback` immediately. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_007: [** If `connect` is called while the transport is waiting between operation status queries, it shall do nothing and call `callback` immediately. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_008: [** If `connect` is called while the transport is executing an operation status query, it shall do nothing and call `callback` immediately. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_009: [** If `connect` is called while the transport is disconnecting, it shall wait for the disconnection to complete, then initiate the connection. **]**
+
+
+### register(registrationId: string, authorization: string | X509, requestBody: any, forceRegistration: boolean, callback: Provisioning.ResponseCallback): void
+Register round-trips one step of the registration process, not returning until the operation is complete
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_010: [** `register` shall `connect` the transport if it is not connected. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_011: [** `register` shall fail if the connection fails. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_012: [** `register` shall call `_doFirstRegistrationRequestForFsm`. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_013: [** If the registration request fails, `register` shall fail. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_014: [** If the registration request returns with status==assigned, it shall call `callback` with null, the response body, and the protocol-specific result. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_015: [** If the registration request returns with status==Assigning, it shall begin polling for operation status requests. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_016: [** If the registration request returns with an unknown status, `register` shall fail with a `SyntaxError` and pass the response body and the protocol-specific result to the `callback`. **]**
```

Comment:
>`SyntaxError` [](start = 149, length = 13)

!

Reply:
agreed. more explicit error objects (like invalid server response) coming in the future.

---
In reply to: [146398661](https://github.com/Azure/azure-iot-sdk-node/pull/115#discussion_r146398661) [](ancestors = 146398661)